File: gersemi/mode.py
from enum import Enum


class Mode(Enum):
    ForwardToStdout = 0
    RewriteInPlace = 1
    CheckFormatting = 2
    ShowDiff = 3
    PrintConfig = 4
    CheckFormattingAndShowDiff = 5


def get_mode(args) -> Mode:
    if args.check_formatting and args.show_diff:
        return Mode.CheckFormattingAndShowDiff
    if args.show_diff:
        return Mode.ShowDiff
    if args.check_formatting:
        return Mode.CheckFormatting
    if args.in_place:
        return Mode.RewriteInPlace
    if args.print_config:
        return Mode.PrintConfig
    return Mode.ForwardToStdout

File: gersemi/rust-backend/mode.rs
use pyo3::exceptions::PyRuntimeError;
use pyo3::types::PyAnyMethods;
use pyo3::{Borrowed, FromPyObject, PyAny, PyErr};

#[derive(Debug)]
pub enum Mode {
    ForwardToStdout,
    RewriteInPlace,
    CheckFormatting,
    ShowDiff,
    PrintConfig,
    CheckFormattingAndShowDiff,
}

impl FromPyObject<'_, '_> for Mode {
    type Error = PyErr;

    fn extract(obj: Borrowed<'_, '_, PyAny>) -> Result<Self, Self::Error> {
        let value: usize = obj.getattr("value")?.extract()?;
        match value {
            0 => Ok(Self::ForwardToStdout),
            1 => Ok(Self::RewriteInPlace),
            2 => Ok(Self::CheckFormatting),
            3 => Ok(Self::ShowDiff),
            4 => Ok(Self::PrintConfig),
            5 => Ok(Self::CheckFormattingAndShowDiff),
            _ => Err(PyRuntimeError::new_err("Invalid Mode")),
        }
    }
}
